### PR TITLE
chore(ci_visibility): add Auto Test Retries to pytest plugin v2

### PR DESCRIPTION
**Pull request by romainkomorndatadog**
Introduces support for Auto Test Retries to the v2 `pytest` plugin.

This retries tests that have failed their first attempt (except for tests that have failed setup or teardown, which are not retried at all). Failed tests that pass retries result in the session passing (with `pytest` exiting with code 0). Test that fail all retries cause the session to fail (exiting with code 1).

If any tests were retried, a section is added to the final summary report, showing the results:
```
======================= Datadog Auto Test Retries ========================
_________________________________ FAILED _________________________________
FAILED tests/test_some_fail.py::test_func_fails_all_retries
_________________________________ PASSED _________________________________
PASSED tests/test_all_pass.py::TestAllPass::test_method_succeeds_on_5th_retry
PASSED tests/test_all_pass.py::TestAllPass::test_method_succeeds_on_1st_retry
PASSED tests/test_all_pass.py::test_func_succeeds_on_3rd
PASSED tests/test_all_pass.py::test_func_succeeds_first_retry
PASSED tests/test_some_fail.py::TestAllPass::test_method_succeeds_on_5th_retry
PASSED tests/test_some_fail.py::TestAllPass::test_method_succeeds_on_1st_retry
PASSED tests/test_some_fail.py::test_func_succeeds_on_3rd
PASSED tests/test_some_fail.py::test_func_succeeds_first_retry
___________________ Datadog Auto Test Retries summary ____________________
         1 failed, 8 passed (total attempts: 26 failed, 8 passed)
==========================================================================
```

Other changes in this PR:
- Some refactoring to DRY `pytest` retry logic 
- Introduction of API to "stash" data on Test Visibility objects 
- Tweak to ATR logic to make successfully-retried test not affect suite-level-or-above status
- Tweak to telemetry to add `is_retry` tag to `test`-level `event_finished` metric 

Finally, no release note is added as the feature is still unreleased.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/pytest/_atr_utils.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            retry_outcome = _get_outcome_from_retry(item, outcomes)
+
+        InternalTest.atr_finish_retry(
+            test_id, retry_num, retry_outcome.status, retry_outcome.skip_reason, retry_outcome.exc_info
+        )
+
+    return InternalTest.atr_get_final_status(test_id)
+
+
+def _atr_write_report_for_status(
```
Comment (vitor-de-araujo) [inline ddtrace/contrib/pytest/_atr_utils.py:99]: Isn't this function (and the next one) literally identical to the one from EFD?

Comment (romainkomorndatadog) [reply ddtrace/contrib/pytest/_atr_utils.py:99]: It is indeed. I think there are a few more opportunities for DRY-oriented refactors. I don't know if it's that important to do right now.

File: tests/contrib/pytest/test_pytest_atr.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    def test_pytest_atr_no_ddtrace_does_not_retry(self):
+        self.testdir.makepyfile(test_pass=_TEST_PASS_CONTENT)
+        self.testdir.makepyfile(test_fail=_TEST_FAIL_CONTENT)
+        self.testdir.makepyfile(test_errors=_TEST_ERRORS_CONTENT)
+        self.testdir.makepyfile(test_pass_on_retries=_TEST_PASS_ON_RETRIES_CONTENT)
+        self.testdir.makepyfile(test_skip=_TEST_SKIP_CONTENT)
+        rec = self.inline_run()
+        rec.assertoutcome(passed=2, failed=6, skipped=2)
+        assert len(self.pop_spans()) == 0
```
Comment (vitor-de-araujo) [inline tests/contrib/pytest/test_pytest_atr.py:113]: It's not clear to me how the span numbers correlate to the tests run, can you clarify?

**Review by vitor-de-araujo — APPROVED** [review]

File: tests/contrib/pytest/test_pytest_atr.py (hunk 2 above)
Comment (romainkomorndatadog) [reply tests/contrib/pytest/test_pytest_atr.py:113]: There's a pair of implicit assumptions here:
- if we did not create spans (eg: the `ddtrace` plugin was not active), then we didn't retry tests
- if we did not retry tests, the outcome should be correct (eg: we run 10 tests total, and the ones that failed did not pass thanks to retries)

TBH, it doesn't add that much on top of the [`test_disabled`](https://github.com/DataDog/dd-trace-py/pull/11241/files/b444a492eb519a3597f99e39426056af0c09ff3e#diff-48a91ef11a9dbcd767fd3050fa39ca48a9bc8454d47d71b2afc35ebc630686f7R186) test it's based off of.

**Pull request merged**